bin/runtime/src/migrations/custom_transaction_payment_migration.rs

```diff
@@ -0,0 +1,84 @@
+use codec::{Decode, Encode, MaxEncodedLen};
+use frame_support::{
+    log,
+    pallet_prelude::{Get, TypeInfo},
+    storage_alias,
+    traits::OnRuntimeUpgrade,
+    RuntimeDebug,
+};
+use pallet_transaction_payment::{Config, NextFeeMultiplier};
+#[cfg(feature = "try-runtime")]
+use sp_std::vec::Vec;
+
+use crate::Weight;
+
+const TARGET: &str = "runtime::transaction_payment::migration";
+
+#[storage_alias]
+type StorageVersion = StorageValue<TransactionPayment, Releases>;
+
+// copied from transaction payment code as it is not pub there
+#[derive(Encode, Decode, Clone, Copy, PartialEq, Eq, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+enum Releases {
+    /// Original version of the pallet.
+    V1Ancient,
+    /// One that bumps the usage to FixedU128 from FixedI128.
+    V2,
+}
+
+/// Custom migrations the transaction payment pallet from V1 to V2 that only bumps StorageVersion to 2
+pub struct BumpTransactionVersionToV2<T>(sp_std::marker::PhantomData<T>);
+
+impl<T: Config> OnRuntimeUpgrade for BumpTransactionVersionToV2<T> {
+    fn on_runtime_upgrade() -> Weight {
+        let version = StorageVersion::get();
+        if version.is_some() && version.unwrap() == Releases::V2 {
+            log::warn!(
+                target: TARGET,
+                "skipping V1Ancient to V2 migration: executed on wrong storage version.\
+				Expected version V1Ancient, found {:?}",
+                version,
+            );
+            return T::DbWeight::get().reads(1);
+        }
+
+        StorageVersion::put(Releases::V2);
+        T::DbWeight::get().reads_writes(1, 1)
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+        assert_eq!(
+            StorageVersion::get(),
+            None,
+            "Can only upgrade from version V1Ancient"
```

Comment:
Pardon my ignorance - but if `StorageVersion::get()` is `Some(Releases::V2)` already, won't this `assert_eq!` panic?

Reply:
Ah, this is feature-gated so wouldn't happen on production chain. Still, is this some kind of "standard" when it comes to writing these things?

Reply:
Yes, the flow is as follows:
1. `pre_upgrade` and `post_upgrade` are for developer and revievers to test changes using Testnet data.
2. During real Testnet chain update, those hooks are not called, onlt `on_runtime_upgrade`.

1, and 2. can happen in a different time, even few weeks apart, and on different chains, so there are not really coupled. Therefore we need to check in 2. their own checks just to be sure